STYLE-GUIDE.md

```diff
@@ -0,0 +1,285 @@
+# Spectrometer style guide
+
+Through a number of pull requests, reviews, and discussions, a codebase style has emerged.  This document is an attempt to capture that, and prevent further discussions that slow down work unnecessarily.
+
+This style guide is forked from Github's [semantic](https://github.com/github/semantic).
+
+## General guidelines
+
+Make your code look like the code around it. Consistency is the name of the game.
+
+Use `fourmolu` for formatting.  Our CI setup enforces that all changes pass a `fourmolu` run with no differences.
+
+Our CI systems ensure that all patches pass `hlint`'s muster. We have our own set of rules in `.hlint.yaml`.
+
+We strongly recommend adding Haddock documentation to any top-level declaration, unless its purpose is immediately apparent from its name.
+Comments should describe the "why", type signatures should describe the "what", and the code should describe the "how".
+
+## Formatting
+
+2 spaces everywhere. Tabs are forbidden. Haskell indentation can be unpredictable, so generally stick with what your editor suggests.
+There is no hard line-length limit, though if you go beyond 110 or 120 you should generally split it up, especially for type signatures.
+
+### Use applicative notation when constructing simple data types
+
+``` haskell
+thing :: Parser Foo
+
+-- Broke:
+thing = do
+  a <- bar
+  b <- baz
+  pure (Foo a b)
+
+-- Woke:
+thing = Foo <$> bar <*> baz
+```
+
+Code written with applicative notation is less powerful than code written in `do`-notation.  As a result, it can be easier to reason about, as long as it's also readable.
+
+Overreliance on applicative notation can create code that is difficult to read. Don't use applicative notation in combination with operator sections. If in doubt, write it with `do` notation and see if it's more immediately comprehensible.
+
+### Use leading commas for records, exports, and lists
+
+Leading commas make it easy to add and remove fields without introducing syntax errors, and properly aligned records are easy to read:
+
+``` haskell
+data Pos = Pos
+  { posLine   :: Int
+  , posColumn :: Int
+  }
+```
+
+This rule is enforced by CI.
+
+## Naming
+
+Locally bound variables (such as the arguments to functions, or helpers defined in a `where` clause) can have short names, such as `x` or `go`. Globally bound functions and variables should have descriptive names.
+
+You'll often find yourself implementing functions that conflict with Prelude definitions. If this is the case, avoid adding a prefix to these functions, and instead import them qualified.
+
+``` haskell
+-- Bad
+foo = heapLookup thing
+-- Better
+foo = Heap.lookup thing
+```
+
+## Functions
+
+### Don't go crazy with infix operators
+
+Sensible use of infix operators can provide serious readability benefits, but often the best tool is just a named function. If you're defining new operators, make sure that you have a solid justification for doing so.
+
+### Avoid list comprehensions
```

Comment:
😢

Reply:
Yeah, I know you like them, but I think they're impossible to read most of the time.  This also comes directly from the original guide from `semantic`.